src/utils/document-revision-comparison-utils.js

```diff
@@ -0,0 +1,211 @@
+  add: 'add',
+  delete: 'delete',
+  modify: 'modify'
+};
+
+function mapSectionChangeTypeToDiffType(sectionChangeType) {
+  switch (sectionChangeType) {
+    case SECTION_CHANGE_TYPE.added:
+      return DIFF_TYPE.add;
+    case SECTION_CHANGE_TYPE.removed:
+      return DIFF_TYPE.delete;
+    default:
+      return DIFF_TYPE.modify;
+  }
+}
+
+export function getInterleavedSectionsChanges(oldSectionKeys, newSectionKeys) {
+  const oldSectionKeySet = new Set(oldSectionKeys);
+  const newSectionKeySet = new Set(newSectionKeys);
+  const orderedCommonKeysInOldSection = oldSectionKeys.filter(value => newSectionKeySet.has(value));
+  const orderedCommonKeysInNewSection = newSectionKeys.filter(value => oldSectionKeySet.has(value));
+  const longestCommonSubsequence = mdiff(orderedCommonKeysInOldSection, orderedCommonKeysInNewSection).getLcs();
+  const nonMovedSectionKeySet = new Set(longestCommonSubsequence);
+  const movedSectionKeySet = new Set(orderedCommonKeysInNewSection.filter(value => !nonMovedSectionKeySet.has(value)));
+
+  const result = [];
+  const processedSectionKeys = new Set();
+  let indexInOldSections = 0;
+  let indexInNewSections = 0;
+
+  while (indexInOldSections < oldSectionKeys.length || indexInNewSections < newSectionKeys.length) {
+    const oldSectionKey = oldSectionKeys[indexInOldSections] ?? null;
+    const newSectionKey = newSectionKeys[indexInNewSections] ?? null;
+
+    if (oldSectionKey === newSectionKey) {
+      // section stayed in the same place
+      result.push([oldSectionKey, SECTION_CHANGE_TYPE.unchanged]);
+      processedSectionKeys.add(oldSectionKey);
+      indexInOldSections += 1;
+      indexInNewSections += 1;
+    } else if (newSectionKey !== null && !oldSectionKeySet.has(newSectionKey)) {
+      // section was added in `sectionKeys2`
+      result.push([newSectionKey, SECTION_CHANGE_TYPE.added]);
+      processedSectionKeys.add(newSectionKey);
+      indexInNewSections += 1;
+    } else if (oldSectionKey !== null && !newSectionKeySet.has(oldSectionKey)) {
+      // section was deleted in `sectionKeys2`
+      result.push([oldSectionKey, SECTION_CHANGE_TYPE.removed]);
+      processedSectionKeys.add(oldSectionKey);
+      indexInOldSections += 1;
+    } else if (newSectionKey !== null && movedSectionKeySet.has(newSectionKey)) {
+      // section was moved to here in `sectionKeys2`
+      result.push([newSectionKey, SECTION_CHANGE_TYPE.movedHere]);
+      processedSectionKeys.add(newSectionKey);
+      indexInNewSections += 1;
+    } else if (oldSectionKey !== null && movedSectionKeySet.has(oldSectionKey)) {
+      // section was moved from here to a different place in `sectionKeys2`
+      result.push([oldSectionKey, processedSectionKeys.has(oldSectionKey) ? SECTION_CHANGE_TYPE.movedUp : SECTION_CHANGE_TYPE.movedDown]);
+      processedSectionKeys.add(oldSectionKey);
+      indexInOldSections += 1;
+    } else {
+      throw new Error('Unexpected case while creating section diff');
```

Comment:
It would introduce yet another case that we have to render somehow and we also have already some decisions to make on the UI. This is the kind of `throw` that should never ever be executed unless the algorithm is wrong. I consider it more for developers to act like a compiler error or a guard in case of later changes. By introducing an `unknown` state we would promote the error to be part of the business logic. A section can either be added, deleted, moved, or stay put where it was (when it comes to placement inside a document). I cannot think of another unknown change type. So I would rather not open this door and have to support it throughout the UI (also: we couldn't even test it as it should never happen). So actually, if anything, I'd rather remove the `else` block completely. Whatcha think?

Reply:
I see your point 🤔 If it were the type of project where error logging were collected and reported to the dev team, this is completely how this would be done, while at the same time having the UI gracefully handle or swallow the scenario.
In our current setup, lots of such throws intended for the dev team are actually seen by the user (like "Was looking for x users but found y") cause we don't have this separation set up. 
So maybe we could combine both worlds a bit and instead of throwing we would console log? This way the user is oblivious to the case, but if they would notice something missing in the comparison and report it, we could ask for a screenshot of the browser console (or see it ourselves when we open the comparison).

Reply:
Good idea. Actually we are neglecting the possibility to log stuff (probably due to the reason you mentioned), so I switched from throwing to logging :+1:

Reply:
Nice 👍